src/common/webrtc/api.rs

```diff
@@ -0,0 +1,463 @@
+        }
+    }
+    pub(crate) async fn next_remote_candidate(&mut self) -> Result<Option<Candidate>, WebRTCError> {
+        match self.signaling_rx.next().await {
+            None => Err(WebRTCError::SignalingDisconnected()),
+            Some(req) => {
+                if let Some(stage) = req.stage {
+                    match stage {
+                        answer_request::Stage::Update(c) => {
+                            if let Some(c) = c.candidate {
+                                log::error!("received candidaate {}", c.candidate);
+                                let c = c.candidate.try_into().ok();
+                                return Ok(c);
+                            } else {
+                                log::error!("received no candidates with this update request");
+                                return Ok(None);
+                            }
+                        }
+                        answer_request::Stage::Error(s) => {
+                            if let Some(status) = s.status {
+                                return Err(WebRTCError::SignalingError(status.message));
+                            }
+                            return Err(WebRTCError::SignalingError("unknown".to_owned()));
+                        }
+                        answer_request::Stage::Done(_) => {
+                            return Ok(None);
+                        }
+                        _ => {
+                            return Err(WebRTCError::SignalingError("unexpected stage".to_owned()))
+                        }
+                    }
+                }
+                Ok(None)
+            }
+        }
+    }
+    pub fn send_done(&mut self, uuid: String) -> Result<(), WebRTCError> {
+        let answer = AnswerResponse {
+            uuid,
+            stage: Some(answer_response::Stage::Done(AnswerResponseDoneStage {})),
+        };
+        match self.signaling_tx.send_message(answer) {
+            Err(_) => Err(WebRTCError::SignalingDisconnected()),
+            Ok(_) => Ok(()),
+        }
+    }
+}
+
+#[cfg(feature = "native")]
+type Executor<'a> = NativeExecutor<'a>;
+#[cfg(feature = "esp32")]
+type Executor<'a> = Esp32Executor<'a>;
+
+pub struct WebRTCApi<'a, S, D> {
+    executor: Executor<'a>,
+    signaling: Option<WebRTCSignaling>,
+    uuid: Option<String>,
+    transport: WebRTCTransport,
+    certificate: Rc<S>,
+    local_creds: ICECredentials,
+    remote_creds: Option<ICECredentials>,
+    local_ip: Ipv4Addr,
+    dtls: Option<D>,
+}
+
+impl<'a, S, D> WebRTCApi<'a, S, D>
+where
+    S: Certificate,
+    D: DtlsConnector,
+{
+    pub(crate) fn new(
+        executor: Executor<'a>,
+        tx_half: GrpcMessageSender<AnswerResponse>,
+        rx_half: GrpcMessageStream<AnswerRequest>,
+        certificate: Rc<S>,
+        local_ip: Ipv4Addr,
+        dtls: D,
+    ) -> Self {
+        let udp = block_on(executor.run(async { UdpSocket::bind("0.0.0.0:61205").await.unwrap() }));
```

Comment:
no, we just bind a spefic port because it makes it easier to follow packet on wireshark. In pratice we could use 0 so random port. Since there is a pretty high guarantee that the port will always be free on the ESP32 I am going yo leave it for this PR but adding a ticket to cover this